Kubelet: Add docker operation timeout and monitor timeout rate in e2e test

Random-Liu opened the pull request:
For #23563.
Based on #24748, only the last 2 commits are new.

This PR:
1) Add timeout for all docker operations.
2) Add docker operation timeout metrics
3) Cleanup kubelet stats and add runtime operation error and timeout rate monitoring.
4) Monitor runtime operation error and timeout rate in kubelet perf.

@yujuhong 
/cc @gmarek Because of the metrics change.
/cc @kubernetes/sig-node

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/24918)

<!-- Reviewable:end -->

gmarek commented:
We don't gate anything on those particular metrics, so LGTM

euank commented on pkg/kubelet/dockertools/kube_docker_client.go:
```diff
@@ -177,9 +198,13 @@ func (d *kubeDockerClient) PullImage(image string, auth dockertypes.AuthConfig,
+	defer cancel()
 	opts.RegistryAuth = base64Auth
-	resp, err := d.client.ImagePull(getDefaultContext(), opts, nil)
+	resp, err := d.client.ImagePull(ctx, image, opts)
```
I'd expect pulling to frequently take longer than 2 minutes.
Maybe a longer timeout?

Random-Liu replied:
The image pulling is slow, but for now we only set timeout for the "connection construct" phase, which should be relatively fast.
After that we'll wait and periodically check the image pulling progress infinitely.

euank replied:
Thanks for explaining. Perhaps the comment on `defaultTimeout`, or a comment here, should say something similar?

Random-Liu replied:
Sure. Thanks~

yujuhong commented on pkg/kubelet/dockertools/kube_docker_client.go:
```diff
@@ -303,6 +396,18 @@ func parseDockerTimestamp(s string) (time.Time, error) {
+
+// contextError checks the context, and returns error if the context is timeout.
+func contextError(ctx context.Context) error {
+	if ctx.Err() == context.DeadlineExceeded {
```
What if the context fails with other errors? What are the potential errors and are we going to ignore them?

Random-Liu replied:
Another error is `var Canceled = errors.New("context canceled")`.
We never cancel a context, so that should not happen. But I agree that we should still expose the other potential context error just in case.

yujuhong commented on test/e2e/framework/kubelet_stats.go:
```diff
@@ -56,89 +55,218 @@ type KubeletMetric struct {
+// If a apiserver client is passed in, the function will try to get kubelet metrics from metrics grabber;
+// or else, the function will try to get kubelet metrics directly from the node.
+func getKubeletMetricsFromNode(c *client.Client, nodeName string) (metrics.KubeletMetrics, error) {
+	if c != nil {
```
nit: reduce the indent by
```
if c == nil {
    return metrics.GrabKubeletMetricsWithoutProxy(nodeName)
}

// The rest of the code
```

Random-Liu replied:
ACK

yujuhong commented on pkg/kubelet/dockertools/kube_docker_client.go:
```diff
@@ -319,7 +433,7 @@ type containerNotFoundError struct {
 
 func (e containerNotFoundError) Error() string {
-	return fmt.Sprintf("Error: No such container: %s", e.ID)
+	return fmt.Sprintf("no such container: %s", e.ID)
```
Let's switch to "%q" to improve clarity

Random-Liu replied:
ACK.

yujuhong commented on pkg/metrics/kubelet_metrics.go:
```diff
@@ -126,6 +130,22 @@ func NewKubeletMetrics() KubeletMetrics {
 }
 
+// GrabKubeletMetricsWithoutProxy retrieve metrics from the kubelet on the given node using a simple GET over http.
+// Currently only used in integration tests.
```
s/integration/e2e

Random-Liu replied:
This function is only used in the integration test. :)
Here is the [original code](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/kubelet_stats.go#L405-L418).

yujuhong replied:
The code you posted was in test/e2e/framework, where is it used in the integration tests?

Random-Liu replied:
It's [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/integration/integration.go#L857-L860) :)

yujuhong replied:
Ah I see. thanks

Random-Liu committed:
Add timeout for all docker operation.

Random-Liu committed:
1) Add docker operation timeout metrics.
2) Cleanup kubelet stats and add runtime operation error and timeout
rate monitoring.
3) Monitor runtime operation error and timeout rate in
kubelet perf.

Random-Liu commented:
@yujuhong Addressed comments. :)

Random-Liu commented:
@k8s-bot test this issue #25278

yujuhong commented:
Looks good overall. Have you tested the timeout manually to see if it works?

Random-Liu commented:
@yujuhong Yeah, I've tested with timeout=100ms and 10ms, it works. :)
